src/pybind/mgr/dashboard/controllers/host.py

```diff
@@ -151,9 +156,25 @@ def merge_hosts_by_hostname(ceph_hosts, orch_hosts):
             }, orch_hosts_map[hostname]) for hostname in orch_hosts_map
     ]
     hosts.extend(orch_hosts_only)
+    for host in hosts:
+        host['service_instances'] = populate_service_instances(
+            host['hostname'], host['services'])
     return hosts
 
 
+def populate_service_instances(hostname, services):
+    orch = OrchClient.instance()
+    if orch.available():
+        return Counter([
+            daemon['daemon_type'] for daemon in [
+                d.to_dict() for d in orch.services.list_daemons(
+                    hostname=hostname)]])
+
+    return Counter([
+        daemon['type'] for daemon in services
+    ])
```

Comment:
Minor thing: if you remove the enclosing square brackets then the list comprehension becomes a generator expression:
```suggestion
        return Counter(
            daemon['daemon_type'] for daemon in [
                d.to_dict() for d in orch.services.list_daemons(
                    hostname=hostname)])

    return Counter(
        daemon['type'] for daemon in services
    )
```

The difference is that list comprehensions generate first a real `list` and then pass that object to the `Counter()` constructor. On the other hand, generators will behave as an iterator, that emits individual items on `next()` call, and allows a consumer to receive them one by one (instead of the whole list object together).

Reply:
updated, thanks